src/node-capnp/capnp.cc

```diff
@@ -463,31 +586,47 @@ class UvIoStream: public OwnedFileDescriptor, public kj::AsyncIoStream {
                                   kj::ArrayPtr<const kj::ArrayPtr<const byte>> morePieces) {
     KJ_STACK_ARRAY(struct iovec, iov, 1 + morePieces.size(), 16, 128);
 
-    // writev() interface is not const-correct.  :(
     iov[0].iov_base = const_cast<byte*>(firstPiece.begin());
     iov[0].iov_len = firstPiece.size();
     for (uint i = 0; i < morePieces.size(); i++) {
       iov[i + 1].iov_base = const_cast<byte*>(morePieces[i].begin());
       iov[i + 1].iov_len = morePieces[i].size();
     }
 
-    ssize_t writeResult;
-    KJ_NONBLOCKING_SYSCALL(writeResult = ::writev(fd, iov.begin(), iov.size())) {
-      // Error.
+    size_t totalBytesWritten = 0;
+    for(uint i = 0; i < morePieces.size() + 1; i++) {
+      struct iovec* thisIov = &iov[i];
+
+      size_t bytesWritten = 0;
+      while (bytesWritten < thisIov->iov_len) {
+        int writeResult;
+        KJ_NONBLOCKING_SYSCALL(writeResult, ::write(
+          fd,
+          (const void*) (((const byte*) thisIov->iov_base) + bytesWritten),
+          thisIov->iov_len - bytesWritten
+        )) {
+          // Error.
+
+          // We can't "return kj::READY_NOW;" inside this block because it causes a memory leak 
+          // due to a bug that exists in both Clang and GCC:
+          //   http://gcc.gnu.org/bugzilla/show_bug.cgi?id=33799
+          //   http://llvm.org/bugs/show_bug.cgi?id=12286
+          goto error;
+        }
+        if (false) {
+        error:
+          return kj::READY_NOW;
+        }
 
-      // We can't "return kj::READY_NOW;" inside this block because it causes a memory leak due to
-      // a bug that exists in both Clang and GCC:
-      //   http://gcc.gnu.org/bugzilla/show_bug.cgi?id=33799
-      //   http://llvm.org/bugs/show_bug.cgi?id=12286
-      goto error;
-    }
-    if (false) {
-    error:
-      return kj::READY_NOW;
+        // A negative result means EAGAIN, which we can treat the same as having written zero bytes.
+        size_t n = writeResult < 0 ? 0 : writeResult;
+        
+        bytesWritten += n;
+        totalBytesWritten += n;
+      }
     }
 
-    // A negative result means EAGAIN, which we can treat the same as having written zero bytes.
-    size_t n = writeResult < 0 ? 0 : writeResult;
+    size_t n = totalBytesWritten;
```

Comment:
This change seems to cause Linux (and I suspect macos) clients to hang their event loop as soon as a capability is used to dispatch an RPC request.

Reply:
(Applying [this commit](https://github.com/fourieraudio/node-capnp/commit/d6580a7009e39756a247b3a8270311c909a79e6c) allows the client to run normally on Linux)

Reply:
@p--b: On the current `win32-cross` branch, would you expect this bug to reproduce if I `cargo run --release` Director, start up the new Transform frontend (the `client-infra` branch of Transform), and connect to `127.0.0.1`?

If so, this bug isn't reproducing on Windows. I don't have a Linux test-case which I can use to investigate that side of it (because our frontend doesn't run on WSL, seemingly due to filesystem permission errors). I could try to fix this by staring at it and thinking hard, but the first five minutes of that haven't led to any epiphanies, except that the `writeResult < 0` comparison looks a bit suspicious.

Do you want to look into this, or should we simply `#ifdef` it and pray for forgiveness later?

Reply:
This should now be patched